Feature/iiwa 54

Xiscoder opened the pull request:
Merge 54 into 24

Commit:
IIWA-35: Add .gitignore for backend-api and Visual Studio.

charkour committed:
IIWA-35: Run Scaffold-DbContext. Note: Make sure Window's User has db ownership.

charkour committed:
IIWA-35: Fixed namespace and registered ITInvetoryDBContext with dependency injection.

charkour committed:
IIWA-35: Added EmployeeController template.

charkour committed:
IIWA-35: Run Scaffold DbContext after updated db modles.

charkour committed:
IIWA-35: Fix the namespace of models and create template EmployeeController.

charkour committed:
IIWA-35: Re-add DI injection to Program. Added port 4001 for https in Startup. Added authorization to Controllers.

charkour committed:
IIWA-35: Add Automatic AD authentication when running IIS app. Add sample string API call but have to fill in proper username and domain for it to work properly.

charkour committed:
IIWA-35: Final (please!) creation of models to fit new db designs.

charkour committed:
IIWA-35: Added OData for powerful querying from front-end calls.

charkour committed:
IIWA-35: Rebuild one model and the context after changing some columns.

Xiscoder committed:
IIWA-54:Built cost controller and build an endpoint within that controller to calculate costs from various tables

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+            _context = context;
+        }
+
+        // GET: api/Cost
```
You should say how the data is returned and in what form it can be expected.

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        // GET: api/Cost
+        [Route("Dashboard")]
```
Would the route be then api/cost/dashboard ? Fix the GET: comment.

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+        // GET: api/Cost
+        [Route("Dashboard")]
+        [HttpGet]
+        [EnableQuery]
```
In the OData video they write `EnableQuery()` like it is a method. Without the parenthesis does it still work?

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+                    var PluginsForThatProgram = PluginsList.Where(x => x.ProgramId == Program.ProgramID);
+                    foreach( var PluginForThatProgram in PluginsForThatProgram)
+                    {
+                        CostOfPluginsPerYear += PluginForThatProgram.PluginCostPerYear;
```
Having visual studio auto format the document would help with make consistent spacing in this file.

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+
+            var DistinctProgramsList = ProgramsList.GroupBy(x => x.ProgramName).Select(x => x.FirstOrDefault());
+            //looping through those distinct programs and if they have plugins, calculating the cost of these plugins
+            foreach (var Program in DistinctProgramsList)
```
Do you need to check if the plugin is deleted?

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+
+            // looping through server table and finding servers that are not "deleted" 
+            // and that are not null. Adding up the cost of those servers
+            foreach (var Server in ServerList)
```
Do they want servers in the program cost?

charkour commented on src/backend-api/Controllers/CostController.cs:
```diff
@@ -0,0 +1,89 @@
+            }
+
+            var DistinctProgramsList1 = ProgramsList.GroupBy(x => x.ProgramName).Select(x => x.FirstOrDefault()).Select(x => x.ProgramName);
+            decimal? monthlyCost = CostOfProgramsPerYear / 12;
```
Is monthly cost being returned?

charkour reviewed (changes requested):
Define the return type and the path in the comments.
Decide if the cost per year and per month need to both be returned. Which I think it should.
Remove the servers from the cost of programs.

Xiscoder committed:
IIWA-54 Added methods in models to allow for accesing of those fields in the databse.
Cleaned up my dashboard cost endpoint.
Dashboard endpoint now only returns programs and their plugins and now returns as an array

charkour reviewed (approved):
Looks good.
